beluga_ros/src/amcl.cpp

```diff
@@ -50,23 +50,47 @@ void Amcl::update_map(beluga_ros::OccupancyGrid map) {
   std::visit([&](auto& sensor_model) { sensor_model.update_map(std::move(map)); }, sensor_model_);
 }
 
+// Overloaded update method for LaserScan.
 auto Amcl::update(Sophus::SE2d base_pose_in_odom, beluga_ros::LaserScan laser_scan)
     -> std::optional<std::pair<Sophus::SE2d, Sophus::Matrix3d>> {
-  if (particles_.empty()) {
-    return std::nullopt;
-  }
-
-  if (!update_policy_(base_pose_in_odom) && !force_update_) {
-    return std::nullopt;
-  }
-
   // TODO(nahuel): Remove this once we update the measurement type.
   auto measurement = laser_scan.points_in_cartesian_coordinates() |  //
                      ranges::views::transform([&laser_scan](const auto& p) {
                        const auto result = laser_scan.origin() * Sophus::Vector3d{p.x(), p.y(), 0};
                        return std::make_pair(result.x(), result.y());
                      }) |
                      ranges::to<std::vector>;
+  return update(base_pose_in_odom, std::move(measurement));
+}
+
+// Overloaded update method for SparsePointCloud3.
+auto Amcl::update(Sophus::SE2d base_pose_in_odom, beluga_ros::SparsePointCloud3f point_cloud)
+    -> std::optional<std::pair<Sophus::SE2d, Sophus::Matrix3d>> {
+  std::vector<std::pair<double, double>> measurement;  // NOTE: Should be float?
+  measurement.reserve(point_cloud.size());
+
+  // Transform points from the sensor frame to the base frame (and project to Z=0).
+  measurement = point_cloud.points() | ranges::views::transform([&point_cloud](const auto& p) {
+                  const auto result = point_cloud.origin() * p.template cast<double>();
+                  return std::pair{result.x(), result.y()};
+                }) |
+                ranges::to<std::vector>();
```

Comment:
Please give lambdas used in ranges it a self-documenting name. The range will be easier to read too.